src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs

```diff
@@ -4400,350 +4405,645 @@ internal static IEnumerable<CompletionResult> CompleteFilename(CompletionContext
+                }
+            }
+
+            return results;
+        }
+
+        private static string GetChildNameFromPsObject(dynamic psObject, char separator)
+        {
+            // The obvious solution would be to use the "PSChildName" property
+            // but some providers don't have it (like the Variable provider)
+            // So we use a substring of "PSPath" instead.
+            string childName = psObject.PSPath ?? string.Empty;
+            int ProviderSeparatorIndex = childName.IndexOf("::", StringComparison.Ordinal);
+            childName = childName.Substring(ProviderSeparatorIndex + 2);
+            int indexOfName = childName.LastIndexOf(separator);
+            if (indexOfName == -1 || indexOfName + 1 == childName.Length)
+            {
+                return childName;
+            }
+
+            return childName.Substring(indexOfName + 1);
+        }
+        
+        /// <summary>
+        /// Takes a path and rebuilds it with the specified variable replacements.
+        /// Also escapes special characters as needed.
+        /// </summary>
+        private static string RebuildPathWithVars(
+            string path,
+            string homePath,
+            StringConstantType stringType,
+            bool literalPath,
+            out bool quotesAreNeeded)
+        {
+            var sb = new StringBuilder(path.Length);
+            int homeIndex = string.IsNullOrEmpty(homePath)
+                ? -1
+                : path.IndexOf(homePath, StringComparison.OrdinalIgnoreCase);
+            quotesAreNeeded = false;
+            bool useSingleQuoteEscapeRules = stringType is StringConstantType.SingleQuoted or StringConstantType.BareWord;
+
+            for (int i = 0; i < path.Length; i++)
+            {
+                if (i == homeIndex)
+                {
+                    _ = sb.Append('~');
+                    i += homePath.Length - 1;
+                    continue;
+                }
+
+                EscapeCharIfNeeded(sb, path, i, stringType, literalPath, useSingleQuoteEscapeRules, ref quotesAreNeeded);
+                _ = sb.Append(path[i]);
+            }
+
+            return sb.ToString();
+        }
+
+        private static string EscapePath(string path, StringConstantType stringType, bool literalPath, out bool quotesAreNeeded)
+        {
+            var sb = new StringBuilder(path.Length);
+            bool useSingleQuoteEscapeRules = stringType is StringConstantType.SingleQuoted or StringConstantType.BareWord;
+            quotesAreNeeded = false;
+
+            for (int i = 0; i < path.Length; i++)
+            {
+                EscapeCharIfNeeded(sb, path, i, stringType, literalPath, useSingleQuoteEscapeRules, ref quotesAreNeeded);
+                _ = sb.Append(path[i]);
+            }
+
+            return sb.ToString();
+        }
+
+        private static void EscapeCharIfNeeded(
```

Comment:
Not for the PR. Maybe useful .Net implementation https://source.dot.net/#System.Text.RegularExpressions/System/Text/RegularExpressions/RegexParser.cs,151